master/internal/stream/system.go

```diff
@@ -201,35 +196,21 @@ func (ps *PublisherSet) Websocket(
 
 	streamer := stream.NewStreamer()
 	user := c.(*detContext.DetContext).MustGetUser()
-	ss, err := NewSubscriptionSet(ctx, streamer, ps, user)
-	if err != nil {
-		return errors.Wrap(err, "creating subscription set")
-	}
-	defer ss.UnsubscribeAll()
-
-	// First read the startup message.
-	var startupMsg StartupMsg
-	err = socket.ReadJSON(&startupMsg)
-	// XXX: errors here don't seem to appear on the websocket side...?
-	if err != nil {
-		return errors.Wrap(err, "reading startup message")
-	}
-	// Use the declarative strategy to process all offline events:
-	//   - insertions
-	//   - updates
-	//   - deletions
-	//   - appearances
-	//   - disappearances
-	//   - fallin
-	//   - fallout
-	msgs, err := ss.Startup(ctx, user, startupMsg)
-	if err != nil {
-		return errors.Wrapf(err, "gathering startup messages")
-	}
-	err = writeAll(socket, msgs)
-	if err != nil {
-		return errors.Wrapf(err, "writing startup messages")
-	}
+	startupErrChan := make(chan error)
+	go func() {
+		var ss SubscriptionSet
+		defer ss.UnsubscribeAll()
+		for {
+			err := startupMsgHandler(ctx, socket, ps, &ss, user, streamer)
```

Comment:
You still have half the business logic in the main websocket thread (the online-streaming logic) and the other half (the startup logic) in this side goroutine.

Can you guarantee that after sending a sync message from the side-goroutine is synchronized with the online-streaming logic?

I _really_ think you should find a way to put all the business logic on the same thread; these sorts of things become trivial to solve.